src/sentry/db/models/base.py

```diff
@@ -92,6 +96,65 @@ class Meta:
     __repr__ = sane_repr("id")
 
 
+@dataclass
+class SnowflakeBitSegment:
+    length: int
+    name: str
+
+    def validate(self, value):
+        if self.length <= 0:
+            raise Exception("The length should be a positive number")
+        if value >> self.length != 0:
+            raise Exception(f"{self.name} exceed max bit value of {self.length}")
+        return True
+
+
+class Snowflake:
+    SENTRY_EPOCH_START = datetime(2022, 4, 26, 0, 0).timestamp()
+    SNOWFLAKE_ID_LENGTH = getattr(settings, "SNOWFLAKE_ID_LENGTH", 53)
+    SNOWFLAKE_VERSION_ID_LENGTH = getattr(settings, "SNOWFLAKE_VERSION_ID_LENGTH", 5)
+    SNOWFLAKE_TIME_DIFFERENCE_LENGTH = getattr(settings, "SNOWFLAKE_TIME_DIFFERENCE_LENGTH", 32)
+    SNOWFLAKE_REGION_ID_LENGTH = getattr(settings, "SNOWFLAKE_REGION_ID_LENGTH", 12)
+    SNOWFLAKE_REGION_SEQUENCE_LENGTH = getattr(settings, "SNOWFLAKE_REGION_SEQUENCE_LENGTH", 4)
+    SNOWFLAKE_ID_VALIDATOR = SnowflakeBitSegment(SNOWFLAKE_ID_LENGTH, "Snowflake ID")
+
+    SEGMENT_LENGTH = OrderedDict()
+    SEGMENT_LENGTH["VERSION_ID"] = SnowflakeBitSegment(SNOWFLAKE_VERSION_ID_LENGTH, "Version ID")
+    SEGMENT_LENGTH["TIME_DIFFERENCE"] = SnowflakeBitSegment(
+        SNOWFLAKE_TIME_DIFFERENCE_LENGTH, "Time difference"
+    )
+    SEGMENT_LENGTH["REGION_ID"] = SnowflakeBitSegment(SNOWFLAKE_REGION_ID_LENGTH, "Region ID")
+    SEGMENT_LENGTH["REGION_SEQUENCE"] = SnowflakeBitSegment(
+        SNOWFLAKE_REGION_SEQUENCE_LENGTH, "Region sequence"
+    )
```

Comment:
I'd like to factor out the string keys, store the `SnowflakeBitSegment` objects as constants, and use those constants themselves as keys to the dict of segment values.

If you change the declaration to
```
@dataclass(frozen=True, eq=True)
class SnowflakeBitSegment:
```
then the objects will be hashable.